packages/cardano-services/src/Asset/TypeormAssetProvider/TypeormAssetProvider.ts

```diff
@@ -58,26 +58,31 @@ export class TypeormAssetProvider extends TypeormProvider implements AssetProvid
       );
     }
 
-    return this.withQueryRunner(async (queryRunner) => {
-      const assetInfoList = await Promise.all(assetIds.map((assetId) => this.#getAssetInfo(assetId, queryRunner)));
-
-      if (extraData?.nftMetadata) {
-        await Promise.all(
-          assetInfoList.map(async (assetInfo) => {
-            assetInfo.nftMetadata = await this.#getNftMetadata(assetInfo, queryRunner);
-          })
-        );
-      }
+    const { nftMetadata, tokenMetadata } = extraData || {};
 
-      if (extraData?.tokenMetadata) {
-        const tokenMetadataList = await this.#fetchTokenMetadataList(assetIds);
+    const [assetInfoMap, metadataMap, tokeMap] = await this.withDataSource(async (dataSource) =>
+      Promise.all([
+        this.#getAssetsInfo(assetIds, dataSource),
+        nftMetadata
+          ? this.#getMultiNftMetadata(assetIds, dataSource)
+          : Promise.resolve(new Map<Cardano.AssetId, Asset.NftMetadata>()),
+        tokenMetadata
+          ? this.#fetchTokenMetadataList(assetIds).then(
+              (list) => new Map(list.map((metadata, id) => [assetIds[id], metadata]))
+            )
+          : Promise.resolve(new Map<Cardano.AssetId, Asset.TokenMetadata | null | undefined>())
+      ])
+    );
 
-        for (const [index, assetInfo] of assetInfoList.entries()) {
-          assetInfo.tokenMetadata = tokenMetadataList[index];
-        }
-      }
+    return assetIds.map((assetId) => {
+      const assetInfo = assetInfoMap.get(assetId);
+
+      if (!assetInfo) throw new ProviderError(ProviderFailure.NotFound, undefined, `Asset not found '${assetId}'`);
+
+      if (nftMetadata) assetInfo.nftMetadata = metadataMap.get(assetId) || null;
+      if (tokenMetadata) assetInfo.tokenMetadata = tokeMap.get(assetId);
```

Comment:
why does this line not need `|| null` but the previous does?

Reply:
Honestly IDK the actual answer to the question.
What I can say is previous version was using directly the result from [`this.#fetchTokenMetadataList(assetIds)`](https://github.com/input-output-hk/cardano-js-sdk/blob/2da68e66341ecffe674009009a11d28d7e4a6d48/packages/cardano-services/src/Asset/TypeormAssetProvider/TypeormAssetProvider.ts#L73-L77) to populate `assetInfo.tokenMetadata` and in case of missing metadata in the DB, [`null`](https://github.com/input-output-hk/cardano-js-sdk/blob/2da68e66341ecffe674009009a11d28d7e4a6d48/packages/cardano-services/src/Asset/TypeOrmNftMetadataService.ts#L30-L32) was used to populate `assetInfo.nftMetadata`.
I didn't focus on this part of the logic, I focused on the core of the change paying attention to not change the produced result.